Idle during long arcs

AnHardt opened the pull request:
Idle during long arcs
to prevent from watchdog resets during high segmented fast arcs.

Includes:   Make arc support (G2/G3) configurable #3747

Fix for:   Crash (Hard Reset) when issuing G2/G3 command #3729

AnHardt committed:
Make arc support (G2/G3) configurable

Saves about 2669 bytes when deactivated. (About 1% for a AT2560, about __4%__ for a AT644!)

jbrazio commented:
Will close #3747, if needed we can cherry-pick from this PR.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -6484,7 +6488,7 @@ void process_next_command() {
 
       // G2, G3
-      #if DISABLED(SCARA)
+      #if ENABLED(ARC_SUPPORT) & DISABLED(SCARA)
```
Oops. Two ampersands would be better here.

AnHardt replied:
Changed.

thinkyhead commented:
It's hard to perceive the changes, because indentation and changes are combined in a single commit. Would it be possible to add an intermediate commit, with the code changes but not the indentation, and then a commit with the indentation afterward? If that sounds tedious and annoying, I don't mind doing it myself. It just makes it much easier to see the smaller changes that were made, especially for future reference…. esp. as I will need to add similar changes to the Bézier curve support.

jbrazio commented on Marlin/Marlin_main.cpp:
```diff
@@ -7514,8 +7514,16 @@ void prepare_move() {
 
     float feed_rate = feedrate * feedrate_multiplier / 60 / 100.0;
 
+    millis_t previous_ms = millis();
```
You can shave this `millis()` call and init the var as 0.

jbrazio replied:
You can shave this `millis()` call and init the var as 0.

AnHardt commented:
https://github.com/MarlinFirmware/Marlin/pull/3747/files?w=1 
https://github.com/MarlinFirmware/Marlin/pull/3751/commits/89d990800b51d72c56ab92d63b63c3c69a9d2055
Or if you want bot commits togeter but without the changes in the white space
https://github.com/MarlinFirmware/Marlin/pull/3751/files?w=1
?w=1 is your friend. :-)

jbrazio commented on Marlin/Marlin_main.cpp:
```diff
@@ -7423,147 +7427,157 @@ void prepare_move() {
-      r_Y = -offset[X_AXIS] * sin_Ti - offset[Y_AXIS] * cos_Ti;
-      count = 0;
-    }
+    millis_t previous_ms = millis();
```
You can shave this `millis()` call and init the var as 0.

AnHardt replied:
Why should i want that.
That would call `idle()` one time at the beginning of the loop, where we don't need it.

jbrazio replied:
It depends on the perspective..
It was a suggestion.

AnHardt committed:
Idle during long arcs

Idle during long arcs
to prevent from watchdog resets during high segmented fast arcs.

thinkyhead commented:
I'm curious. Why do we disallow arc support with `SCARA`? And it looks like actually we should add a sanity-check for `ARC_SUPPORT` plus `SCARA` now, if they are indeed incompatible.
